src/stream/src/executor/flow_control.rs

```diff
@@ -32,37 +32,42 @@ use super::*;
 /// It is used to throttle problematic MVs that are consuming too much resources.
 pub struct FlowControlExecutor {
     input: BoxedExecutor,
-    rate_limit: u32,
+    rate_limit: Option<u32>,
 }
 
 impl FlowControlExecutor {
     #[allow(clippy::too_many_arguments)]
-    pub fn new(input: Box<dyn Executor>, rate_limit: u32) -> Self {
+    pub fn new(input: Box<dyn Executor>, rate_limit: Option<u32>) -> Self {
         #[cfg(madsim)]
         println!("FlowControlExecutor rate limiter is disabled in madsim as it will spawn system threads");
         Self { input, rate_limit }
     }
 
     #[try_stream(ok = Message, error = StreamExecutorError)]
     async fn execute_inner(self) {
-        let quota = Quota::per_second(NonZeroU32::new(self.rate_limit).unwrap());
-        let clock = MonotonicClock;
-        let rate_limiter = RateLimiter::direct_with_clock(quota, &clock);
+        let get_rate_limiter = |rate_limit: u32| {
+            let quota = Quota::per_second(NonZeroU32::new(rate_limit).unwrap());
+            let clock = MonotonicClock;
+            RateLimiter::direct_with_clock(quota, &clock)
+        };
+        let rate_limiter = self.rate_limit.map(get_rate_limiter);
         #[for_await]
         for msg in self.input.execute() {
             let msg = msg?;
             match msg {
                 Message::Chunk(chunk) => {
                     #[cfg(not(madsim))]
                     {
-                        let result = rate_limiter
-                            .until_n_ready(NonZeroU32::new(chunk.cardinality() as u32).unwrap())
-                            .await;
-                        if let Err(InsufficientCapacity(n)) = result {
-                            tracing::error!(
-                                "Rate Limit {} smaller than chunk cardinality {n}",
-                                self.rate_limit,
-                            );
+                        if let Some(rate_limiter) = &rate_limiter {
```

Comment:
Can we move this outside of the loop? Otherwise when processing each chunk we will incur a branch.

Reply:
I am planning to impl config change here.
There will be another match arm handling rate limit changes

```rust
match msg {
  Message::Chunk(chunk) => { .. } // run normally
  Message::Barrier(barrier) => {
    let new_rate_limit = extreact_rate_limit(barrier);
    if new_rate_limit == 0 { rate_limiter = None; } 
    else { rate_limiter = get_rate_limiter(new_rate_limit); }
  }
}
```

Reply:
Hmm I still think even with that, we can avoid the branch in the inner loop. But I won't block this PR on that.
Will see if I can refactor it after the work is done for it.

Reply:
https://github.com/risingwavelabs/risingwave/issues/13062

Reply:
> Hmm I still think even with that, we can avoid the branch in the inner loop

Seems inevitable because separate logic for data chunk and barrier. Looking forward to your solution!